tensorpack/tfutils/symbolic_functions.py

```diff
@@ -212,3 +212,149 @@ def saliency_map(output, input, name="saliency_map"):
     saliency_op = tf.gradients(max_outp, input)[:][0]
     saliency_op = tf.identity(saliency_op, name=name)
     return saliency_op
+
+
+def contrastive_loss(left, right, y, margin, extra=False):
+    """Loss for Siamese networks as described in the paper:
+    `Learning a Similarity Metric Discriminatively, with Application to Face
+    Verification <http://yann.lecun.com/exdb/publis/pdf/chopra-05.pdf>` by Chopra et al.
+
+    .. math::
+        \min 0.5 * Y * D^2 + 0.5 * (1-Y) * \{\max(0, m - D)\}^2, D = \Vert l - r \Vert_2
+
+    Args:
+        left (tf.Tensor): left of image pair
+        right (tf.Tensor): right of image pair
+        y (tf.Tensor): label (1: similar, 0:not similar)
+        margin (float): horizont for negative examples (y==0)
+        extra (bool, optional): return distances for pos and negs
+
+    Returns:
+        tf.Tensor: constrastive_loss, [pos_dist, neg_dist]
+    """
+    with tf.name_scope("constrastive_loss"):
+        y = tf.cast(y, tf.float32)
+
+        diff = left - right
+
+        delta = tf.reduce_sum(tf.square(diff), 1)
+        delta_sqrt = tf.sqrt(delta + 1e-6)
+
+        match_loss = delta
+        missmatch_loss = tf.square(tf.nn.relu(margin - delta_sqrt))
+
+        loss = tf.reduce_mean(0.5 * (y * match_loss + (1 - y) * missmatch_loss))
+        pos_dist = tf.reduce_mean(y * delta_sqrt, name="pos-dist")
+        neg_dist = tf.reduce_mean((1 - y) * delta_sqrt, name="neg-dist")
```

Comment:
Is it a good idea to use ``reduce_mean(y * dist)``? Wouldn't it be better to use something like ``reduce_sum(y * dist) / tf.count_nonzero(y)``?
Might not make a big difference when 0/1 labels are perfectly balanced, though.